File: cpp/src/0061.cpp
/*
[61. Rotate List (Medium)](https://leetcode.com/problems/rotate-list/)

@topic: Linked List
@topic: Two Pointers

Constraints:
  - The number of nodes in the list is in the range `[0, 500]`.
  - `-100 <= Node.val <= 100`
  - `0 <= k <= 2 * 10^9`
*/

#include "structs/LinkedList.hpp"

class Solution {
   public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (!head || !head->next || k == 0) return head;
        auto size = 1;
        auto curr = head;
        while (curr->next) {
            ++size;
            curr = curr->next;
        }
        int step = size - k % size;
        if (step == size) {
            return head;
        }
        curr->next = head;
        while (step--) {
            curr = curr->next;
        }
        head = curr->next;
        curr->next = nullptr;
        return head;
    }
};

#include <cassert>
#include <chrono>
#include <iostream>
#include <vector>

using std::tuple;
using std::vector;
using std::chrono::duration_cast;
using std::chrono::microseconds;
using std::chrono::system_clock;

int main() {
    using namespace LinkedList;

    Solution o;

    vector<tuple<ListNode*, int, ListNode*>> CASES = {
        {of({1, 2, 3, 4, 5}), 2, of({4, 5, 1, 2, 3})},
        {of({0, 1, 2}), 4, of({2, 0, 1})},
    };

    for (auto& [head, k, excepted] : CASES) {
        assert(eq(o.rotateRight(head, k), excepted));
    }

    auto start = system_clock::now();
    for (auto i = 0; i < 100000; ++i) {
        o.rotateRight(of({1, 2, 3, 4, 5}), 2);
        o.rotateRight(of({0, 1, 2}), 4);
    }
    auto end = system_clock::now();
    std::cout << duration_cast<microseconds>(end - start).count() << std::endl;
}

File: python/src/0061.py
from typing import Optional

from structs.ListNode import ListNode, create


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head or not head.next or not k:
            return head
        size, walk = 1, head
        while walk.next:
            walk = walk.next
            size += 1
        if (step := size - k % size) == size:
            return head
        walk.next = head
        while step:
            walk = walk.next
            step -= 1
        head = walk.next
        walk.next = None
        return head


if __name__ == '__main__':
    import unittest
    from timeit import timeit

    CASES = (
        (create(1, 2, 3, 4, 5), 2 + 5, create(4, 5, 1, 2, 3)),
        (create(0, 1, 2), 4, create(2, 0, 1)),
    )

    class Test(unittest.TestCase):

        def test(self):
            rotateRight = Solution().rotateRight
            for head, k, excepted in CASES:
                actual = rotateRight(head, k)
                self.assertEqual(actual, excepted)

    unittest.main()

    rotateRight = Solution().rotateRight
    for head, k, _ in CASES:
        print(timeit(lambda: rotateRight(head, k), number=1000000))
